### PR TITLE
big refactor to Stepper class

### PR DESCRIPTION
**Pull request by douglascamata**
We now have a:

- StepWalker class, responsible for walking a small distance towards a
final destination.

- SpiralNavigator class, which defines a route around an area in a spiral. It
determines points to walk to and uses to StepWalker to slowly go from
one point to another.

This architecture enables us to have different Navigators, although it
doesn’t implement a nice Navigator configuration YET.

Implementing https://github.com/PokemonGoF/PokemonGo-Bot/issues/717

### CODE REVIEW

**Comment by jtdroste** [comment]
Going to have to rebase and update your branch my friend :).  Out of date with dev

File: pokemongo_bot/__init__.py [hunk 1]
```diff
@@ -30,11 +32,41 @@ def __init__(self, config):
...
-        self.stepper = Stepper(self)
+        self.step_walker = StepWalker(self)
+        self.navigator = SpiralNavigator(self)
         random.seed()
 
     def take_step(self):
-        self.stepper.take_step()
+        location = self.navigator.take_step()
+        cells = self.find_close_cells(*location)
+        self.work_on_cell(cells[0], location, False)
```
Comment (TheSavior) [inline pokemongo_bot/__init__.py]: What's the plan to have this look at more than just the first cell? Could we change `find_close_cells` to `get_current_metacell` or something?

File: pokemongo_bot/spiral_navigator.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+                        self.y * 0.0025 + self.origin_lon, 0)
+            if self.config.walk > 0:
+                self.bot.step_walker.step(self.config.walk, *position[0:2])
+            else:
+                self.api.set_position(*position)
+        if self.x == self.y or self.x < 0 and self.x == -self.y or self.x > 0 and self.x == 1 - self.y:
+            (self.dx, self.dy) = (-self.dy, self.dx)
+
+        (self.x, self.y) = (self.x + self.dx, self.y + self.dy)
+        sleep(10)
```
Comment (TheSavior) [inline pokemongo_bot/spiral_navigator.py]: Do you have any context in to why we sleep for so long here?

**Comment by solderzzc** [comment]
👍

**Comment by TheSavior** [comment]
This looks right to me except for my comments

**Comment by douglascamata** [comment]
@jtdroste yeah, I will. But I'm afraid of ugly merge conflict, so I'm taking a deep breath first.

**Comment by Nihisil** [comment]
👍

**Comment by douglascamata** [comment]
Don't merge, I found problem.

**Comment by douglascamata** [comment]
@jtdroste need help! My code is not doing small steps like I want it to, but I'm not a geometry guy and you are! :(

**Comment by TheSavior** [comment]
Fixed the merge conflicts. Fixed the branch to successfully catch pokemon, spin forts, and walk around. It also doesn't spam the log with walking info.

**Comment by solderzzc** [comment]
👍

**Comment by solderzzc** [comment]
@tejado  how do you think about this commit? Should we merge?

**Comment by solderzzc** [comment]
`Traceback (most recent call last):
  File "pokecli.py", line 222, in <module>
    main()
  File "pokecli.py", line 213, in main
    bot.take_step()
  File "pokemongo_bot/__init__.py", line 44, in take_step
    self.work_on_cell(cell, location)
  File "pokemongo_bot/__init__.py", line 133, in work_on_cell
    hack_chain = worker.work()
  File "pokemongo_bot/cell_workers/seen_fort_worker.py", line 109, in work
    raise RuntimeError(message)
RuntimeError: Stopped at Pokestop and did not find experience, items or information about the stop cooldown. You are probably softbanned. Try to play on your phone, if pokemons always ran away and you find nothing in PokeStops you are indeed softbanned. Please try again in a few hours.`

Works well in dev branch.

File: pokemongo_bot/__init__.py [hunk 3]
```diff
@@ -78,7 +112,7 @@ def work_on_cell(self, cell, position, include_fort_on_path):
                 if self.catch_pokemon(pokemon) == PokemonCatchWorker.NO_POKEBALLS:
                     break
         if (self.config.mode == "all" or
-                self.config.mode == "farm") and include_fort_on_path:
+                self.config.mode == "farm"):
             if 'forts' in cell:
                 # Only include those with a lat/long
                 forts = [fort
```
Comment (ProjectBarks) [inline pokemongo_bot/__init__.py]: Am i missing something or is this missing a ]

File: pokemongo_bot/spiral_navigator.py (hunk 2 above)
Comment (ProjectBarks) [inline pokemongo_bot/spiral_navigator.py]: @douglascamata

File: pokemongo_bot/step_walker.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+    def __init__(self, bot):
+        self.bot = bot
+        self.api = bot.api
+
+    def step(self, speed, lat, lng):
+        if self.api._position_lat == lat and self.api._position_lng == lng:
+            return True
+
+        dist = distance(
```
Comment (ProjectBarks) [inline pokemongo_bot/step_walker.py]: You dont use the dist variable

File: pokemongo_bot/__init__.py (hunk 3 above)
Comment (TheSavior) [inline pokemongo_bot/__init__.py]: It's on the next line:

```
forts = [fort
                         for fort in cell['forts']
                         if 'latitude' in fort and 'type' in fort]
```

File: pokemongo_bot/step_walker.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+            i2f(self.api._position_lat),
+            i2f(self.api._position_lng),
+            lat,
+            lng
+        )
+
+        dLat = (lat - i2f(self.api._position_lat))
+        dLng = (lng - i2f(self.api._position_lng))
```
Comment (ProjectBarks) [inline pokemongo_bot/step_walker.py]: Just change this to
(dLat, dLng) = ((lat - i2f(self.api._position_lat)), (lng - i2f(self.api._position_lng)))`

File: pokemongo_bot/step_walker.py (hunk 4 above)
Comment (TheSavior) [inline pokemongo_bot/step_walker.py]: Interesting that the linter didn't fail on this.

File: pokemongo_bot/step_walker.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import logger
+
+from cell_workers.utils import distance, i2f, format_time
+from human_behaviour import random_lat_long_delta, sleep
+from math import ceil
```
Comment (ProjectBarks) [inline pokemongo_bot/step_walker.py]: You dont use this import

File: pokemongo_bot/spiral_navigator.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+                    self.steplimit2, self.x, self.y, self.pos, self.dx,
+                    self.dy))
+        # Scan location math
+
+        if -self.steplimit2 / 2 < self.x <= self.steplimit2 / 2 and -self.steplimit2 / 2 < self.y <= self.steplimit2 / 2:
+            position = (self.x * 0.0025 + self.origin_lat,
+                        self.y * 0.0025 + self.origin_lon, 0)
+            if self.config.walk > 0:
+                logger.log('[#] Walking from ' + str((i2f(self.api._position_lat), i2f(
+                    self.api._position_lng))) + " to " + str((str(position[0:2]))))
```
Comment (ProjectBarks) [inline pokemongo_bot/spiral_navigator.py]: Is there a reason we do this `str((str(position[0:2]))))` Why make it a string twice

File: pokemongo_bot/spiral_navigator.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+                'steplimit: {} x: {} y: {} pos: {} dx: {} dy {}'.format(
+                    self.steplimit2, self.x, self.y, self.pos, self.dx,
+                    self.dy))
+        # Scan location math
+
+        if -self.steplimit2 / 2 < self.x <= self.steplimit2 / 2 and -self.steplimit2 / 2 < self.y <= self.steplimit2 / 2:
+            position = (self.x * 0.0025 + self.origin_lat,
+                        self.y * 0.0025 + self.origin_lon, 0)
+            if self.config.walk > 0:
+                logger.log('[#] Walking from ' + str((i2f(self.api._position_lat), i2f(
```
Comment (ProjectBarks) [inline pokemongo_bot/spiral_navigator.py]: use the format instead of +

File: pokemongo_bot/spiral_navigator.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
+# -*- coding: utf-8 -*-
+
+import os
+import json
+import time
+import pprint
```
Comment (ProjectBarks) [inline pokemongo_bot/spiral_navigator.py]: We dont use this import

File: pokemongo_bot/__init__.py [hunk 10]
```diff
@@ -30,13 +32,45 @@ def __init__(self, config):
...
+        )
+        response_dict = self.api.call()
+        map_objects = response_dict.get('responses', {}).get('GET_MAP_OBJECTS', {})
+        status = map_objects.get('status', None)
+
+        map_cells = []
+        if status and status == 1:
+            map_cells = map_objects['map_cells']
+            position = (lat, lng, 0)
+            map_cells.sort(
```
Comment (ProjectBarks) [inline pokemongo_bot/__init__.py]: I may be wrong but you dont need to do key=

**Comment by ProjectBarks** [comment]
@douglascamata I made a lot of comments but make sure you dont have two things primarily
- Unused Variables
- Unused Imports

File: pokemongo_bot/spiral_navigator.py (hunk 7 above)
Comment (TheSavior) [inline pokemongo_bot/spiral_navigator.py]: I believe it does the `(num1, num2)` output.

File: pokemongo_bot/__init__.py [hunk 11]
```diff
@@ -30,13 +32,45 @@ def __init__(self, config):
...
+            longitude=f2i(lng),
+            since_timestamp_ms=timestamp,
+            cell_id=cellid
+        )
+        response_dict = self.api.call()
+        map_objects = response_dict.get('responses', {}).get('GET_MAP_OBJECTS', {})
+        status = map_objects.get('status', None)
+
+        map_cells = []
+        if status and status == 1:
```
Comment (ProjectBarks) [inline pokemongo_bot/__init__.py]: just do `map_objects.get('status')` and `  if status == 1:

File: pokemongo_bot/spiral_navigator.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+            if self.config.walk > 0:
+
+                dist = distance(
+                    i2f(self.api._position_lat),
+                    i2f(self.api._position_lng),
+                    position[0],
+                    position[1]
+                )
+
+                logger.log('[#] Walking from ' + str((i2f(self.api._position_lat), i2f(
```
Comment (ProjectBarks) [inline pokemongo_bot/spiral_navigator.py]: Can we continue on using the .format convention?

File: pokemongo_bot/spiral_navigator.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+                    self.api._position_lng))) + " to " + str((str(position[0:2]))) + " " + format_dist(dist, self.config.distance_unit))
```
Comment (ProjectBarks) [inline pokemongo_bot/spiral_navigator.py]: Why do we do str twice `str((str(position[0:2])))`

File: pokemongo_bot/spiral_navigator.py [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+                    self.steplimit2, self.x, self.y, self.pos, self.dx,
+                    self.dy))
+        # Scan location math
+
+        if -self.steplimit2 / 2 < self.x <= self.steplimit2 / 2 and -self.steplimit2 / 2 < self.y <= self.steplimit2 / 2:
+            position = (self.x * 0.0025 + self.origin_lat,
+                        self.y * 0.0025 + self.origin_lon, 0)
+            if self.config.walk > 0:
+
+                dist = distance(
```
Comment (ProjectBarks) [inline pokemongo_bot/spiral_navigator.py]: We dont use the distance variable

**Comment by MikeDX** [comment]
Looking forward to seeing this merged so I can have a go at implementing tsp for pokestops within the walk radius

**Comment by TheSavior** [comment]
This seems to be structurally working now. I've been running through a bunch of the situations locally and it's been working great.

**Comment by TheSavior** [comment]
@ProjectBarks I'd like to fix most of the style errors for the stepper refactor in a follow up PR where we fix travis and can actually fix all these issues. And get the change in so we don't block all the other stepper changes that want to happen

**Comment by th3w4y** [comment]
I will need to rewrite some of the code that was merge in  #721 because of this refactoring.

I had  **class PolylineStepper(Stepper)**: in polyline_stepper.py I need to rewrite it to a navigator.

**Comment by tejado** [comment]
@solderzzc do you mean me?

**Pull request merged**